src/com/stanimiriliev/chaptersix/pagebean/PageBean.java

```diff
@@ -0,0 +1,143 @@
+
+    private final List<T> src;
+    public final int pageSize;
+    private int upIndex = -1;
+    private int downIndex;
+    private int pageIndex = -1;
+    private boolean hasNext = true, hasPrevious = false;
+    private List<T> list;
+
+    public PageBean(List<T> src, int pageSize){
+        if(src == null || src.size() == 0 || pageSize <= 0)
+            throw new IllegalArgumentException("This error may occur because of several things:\n" +
+                    "Source list is null\nSource list is empty\nPage size is less than or equal to zero");
+        this.src = src;
+        this.pageSize = pageSize;
+    }
+
+    /**
+     * Check if the downIndex or upIndex has reached the its limits
+     * and updates hasPrevious and hasNext variables
+     */
+    private void check(){
+        if(downIndex == 0)
+            hasPrevious = false;
+        if(upIndex == src.size() - 1)
+            hasNext = false;
+        if(upIndex < src.size() - 1 && downIndex > 0){
+            hasNext = true;
+            hasPrevious = true;
+        }
+    }
+
+    /**
+     * Reverses order of the elements of the list
+     * @param list -> The list which will be reversed
+     * @return reversed list
+     */
+    private List<T> reverseList(List<T> list){
+        List<T> temp = new ArrayList<>();
+        for(int i = list.size() - 1; i >= 0; i--)
+            temp.add(list.get(i));
+        return temp;
+    }
+
+    /**
+     * Go to the next page
+     * @return list with the next n elements of the source list (n -> page size)
+     */
+    public List<T> next(){
+        if(hasNext){
+            list = new ArrayList<>();
+            for(int i = downIndex = upIndex + 1; i < src.size() && i < upIndex + 1 + pageSize; i++)
+                list.add(src.get(i));
+            upIndex = downIndex + list.size() - 1;
+            check();
+            pageIndex++;
+            return list;
+        }
+        System.out.println("There is no next page!");
+        return new ArrayList<T>();
+    }
+
+    /**
+     * Go to the previous page
+     * @return list with the previous n elements of the source list (n -> page size)
+     */
+    public List<T> previous(){
+        if(this.hasPrevious){
+            this.list = new ArrayList<>();
+            for(int i = upIndex = downIndex - 1; i >= 0 && i > downIndex - 1 - pageSize; i--)
+                list.add(src.get(i));
+            downIndex = upIndex - list.size() + 1;
+            check();
+            pageIndex--;
+            return reverseList(list);
+        }
+        System.out.println("There is no previous page!");
+        return new ArrayList<T>();
+    }
+
+    public boolean hasNext(){
+        return hasNext;
+    }
+
+    public boolean hasPrevious(){
+        return hasPrevious;
+    }
+
+    /**
+     * Returns first page and make it current page
+     * @return list with the first n elements of the source list (n -> page size)
+     */
+    public List<T> firstPage(){
```

Comment:
could be : 

```
currentPage = 1;///or 0 no so important
 return next();
```
something like that can be applied and for the lastPage : 
calcualting the max page count and check if we have not full page at the end and then we can set up the
currentPage and call next().
Now you it is very hard this to be done, because you have to update too many fields and then to call next()